core/ir/synthetic/encode.c

```diff
@@ -0,0 +1,129 @@
+/* **********************************************************
+ * Copyright (c) 2024 Google, Inc.  All rights reserved.
+ * **********************************************************/
+
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of VMware, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* encode.c -- an encoder for DR synthetic IR */
+
+#include "encode.h"
+
+#include "../globals.h"
+#include "encode_api.h"
+#include "encoding_common.h"
+#include "instr_api.h"
+#include "opnd_api.h"
+
+/* Encodes DR instruction representation \p instr into raw bytes \p encoded_instr.
+ * Returns next instruction's pc.
+ * The encoding scheme followed is described in core/ir/synthetic/encoding_common.h.
+ */
+byte *
+encode_to_synth(dcontext_t *dcontext, instr_t *instr, byte *encoded_instr)
+{
+    /* Use a local uint variable for easier setting of category, eflags, #src, and #dst
+     * values.
+     */
+    uint encoding_header = 0;
+
+    /* Encode number of register destination operands (i.e., written registers).
+     */
+    uint num_dsts = (uint)instr_num_dsts(instr);
+    encoding_header |= num_dsts;
+
+    /* Encode number of register source operands (i.e., read registers).
+     */
+    uint num_srcs = (uint)instr_num_srcs(instr);
+    encoding_header |= (num_srcs << SRC_OPND_SHIFT);
+
+    /* Encode arithmetic flags.
+     */
+    uint eflags_instr = instr_get_arith_flags(instr, DR_QUERY_DEFAULT);
```

Comment:
Can we assert INSTR_EFLAGS_VALID is set?

Reply:
Yes, done.